### PR TITLE
Z Algo

### PR DESCRIPTION
**Pull request by ghost**
### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

### Description

I have added my own code in python for implementing Z algorithm . Z Algorithm searches a sub-string in a string and gives the output as the position of occurrence .

### Checklist

- [x] I've read the contribution guidelines.
- [x] I've checked the issue list before deciding what to submit.
- [x] I've edited the `README.md` and link to my code.

## Related Issues or Pull Requests

#417  (Pull request to solve this issue)
#516  ( I have added totally personalised code in this pull request @SKAUL05 . Last Pull request was not okay as I took help from [here](https://www.tutorialspoint.com/Z-Algorithm),. The site did help me give insights about the problem as I thought I had to do it that way only , and however I wrote the code myself though.. I will be more careful next time 😄 )

### CODE REVIEW

**Comment by ghost** [comment]
@HarshCasper @Kajol-Kumari @SKAUL05 @atarax665  Please do review !

**Review by SKAUL05 — CHANGES_REQUESTED** [review]
- Use meaningful variable names instead of alphabets
- Use docstrings to document functions
- Follow PEP8 guidelines closely

**Comment by ghost** [comment]
I have made the changes in the code , Please review @SKAUL05

File: Python/search/Z_Algorithm_HakinCode.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+"""If list not empty , print positions , else prompt -- pattern not found"""
+if len(positions) != 0:
+
+    """Calling Z__Algo to seacrh the substring positions"""
+    for position in positions:
+        print("The pattern is present at : " + str(position))
+
+else:
+    print("Pattern Not Found !!!")
```
Comment (atarax665) [inline Python/search/Z_Algorithm_HakinCode.py]: - Add sample input/output in the end.
- There's no need to add 'Hakincode' in the file name. Z_Algorithm.py would be fine.

**Review by atarax665 — CHANGES_REQUESTED** [review]

File: Python/search/Z_Algorithm_HakinCode.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
+# Z ALgorithm Function
+def Z__Algo(text, pattern):
+    List = []
+    substring = ""
+
+    """ Iterating upto the index where the substring can occur """
+    for i in range(0, len(text) - len(pattern)+1):
```
Comment (SKAUL05) [inline Python/search/Z_Algorithm_HakinCode.py]: Use meaningful names instead of i

File: Python/search/Z_Algorithm_HakinCode.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+                c += 1
+
+        """ If all the charecters of substring and pattern matches , add position to list """
+        if c == len(pattern):
+            List.append(i)
+
+    """ Return the list that contains all the positions the substring is present at """
+    return List
+
+
```
Comment (SKAUL05) [inline Python/search/Z_Algorithm_HakinCode.py]: Include main and enclose Algo inside a function and call it from main

**Review by SKAUL05 — CHANGES_REQUESTED** [review]

**Comment by ghost** [comment]
I have made further changes according to your requests . Please review @SKAUL05 @atarax665

**Review by SKAUL05 — APPROVED** [review]

**Review by siddharth25pandey — APPROVED** [review]
@SayaniG371 Looks fine to me
Able to merge.
Thanks

**Comment by siddharth25pandey** [comment]
@atarax665 do the review & merge the PR

**Comment by ghost** [comment]
@atarax665 @siddharth25pandey @SKAUL05  I have resolved the merge conflicts, please review

**Comment by ghost** [comment]
@atarax665 Please review the pr, I have resolved the conflicts

**Review by Kajol-Kumari — APPROVED** [review]
LGTM :+1:

**Comment by siddharth25pandey** [comment]
@all-contributors please add @SayaniG371  for her code.

**Pull request merged**